source/agora/node/admin/AdminInterface.d

```diff
@@ -68,47 +74,225 @@ public class AdminInterface
+        used as the get query.
 
-        Begins asynchronous tasks for admin interface
+        Convert LoginInfo to JSON format and respond with the QR code
+        of SVG format.
+        All public key cryptography is done on Curve25519.
+        The signature of the vectorCard provides verification function in the SDK.
+        This LoginInfo has an expiration time.
+        It provide a randomly generated temporary private key encoded
+        in strings on demand.
+        In the future, there is a need to enter a temporary key.
 
     ***************************************************************************/
 
-    public HTTPListener start ()
+    public override string loginQR (out string contentType, out string vary)
     {
-        if (!this.config.admin.enabled)
-            assert(0, "Admin interface is not enabled in config settings.");
+        // The randomly generated temporary KeyPair
+        KeyPair temp_kp = KeyPair.random();
+
+        TimePoint current_time = this.clock.networkTime();
+        // A 'nonce' used to allow expiring
+        current_time += 60 * 60 * 24 * 90; // (default: +90 days)
+        SysTime expires_time = SysTime(unixTimeToStdTime(current_time), UTC());
+
+        LoginInfo login_info = LoginInfo(
+            temp_kp.secret.toString(PrintMode.Clear),
+            VoterCard(
+                this.key_pair.address,
+                temp_kp.address,
+                expires_time.toISOExtString()
+            )
+        );
+
+        login_info.voter_card.signature =
+            this.key_pair.sign(hashFull(login_info.voter_card)[]);
+
+        contentType = "image/svg+xml";
+        vary = "Accept-Encoding";
+        return this.createQRcode(login_info);
+    }
 
-        auto settings = new HTTPServerSettings(this.config.admin.address);
-        settings.port = this.config.admin.port;
+    /***************************************************************************
 
-        auto router = new URLRouter();
+        Respond with QR code containing EncryptionKey.
 
-        // API endpoints for this validator
-        if (this.config.validator.enabled)
-        {
-            auto qrCode = new QRCodeInterface(this.key_pair, this.clock);
+        Enter the name of the APP & height where the encryptionKey will be used
+        as the get query.
 
-            router.get("/loginQR", &qrCode.loginQRCreate);
-            router.get("/encryptionKeyQR", &qrCode.keyQRCreate);
-        }
+    ***************************************************************************/
 
-        return listenHTTP(settings, router);
+    public override string encryptionKeyQR (
+        string app, ulong height, out string contentType, out string vary)
+    {
+        // TODO: PreImage to be released at that height
+        // Or it can be pre-released, so it can be `height + PreimageRevealPeriod`
+        PreImageInfo pre_image;
+
+        // blake2b hashing
+        const Hash key = hashMulti(pre_image, app);
```

Comment:
Can it be pre_image.hash ?

Reply:
I just moved the code. It always uses the default-initialized pre-image so it's not valid anyway...

Reply:
This should be fixed to a block height-based pre-image later.